terror_error.go

```diff
@@ -0,0 +1,262 @@
+func (e *Error) MessageTemplate() string {
+	return e.message
+}
+
+// Error implements error interface.
+func (e *Error) Error() string {
+	describe := e.codeText
+	if len(describe) == 0 {
+		describe = ErrCodeText(strconv.Itoa(int(e.code)))
+	}
+	return fmt.Sprintf("[%s:%s] %s", e.class, describe, e.getMsg())
+}
+
+func (e *Error) getMsg() string {
+	if len(e.args) > 0 {
+		return fmt.Sprintf(e.message, e.args...)
+	}
+	return e.message
+}
+
+// GenWithStack generates a new *Error with the same class and code, and a new formatted message.
+func (e *Error) GenWithStack(format string, args ...interface{}) error {
+	err := *e
+	err.message = format
+	err.args = args
+	return AddStack(&err)
+}
+
+// GenWithStackByArgs generates a new *Error with the same class and code, and new arguments.
+func (e *Error) GenWithStackByArgs(args ...interface{}) error {
+	err := *e
+	err.args = args
+	return AddStack(&err)
+}
+
+// FastGen generates a new *Error with the same class and code, and a new formatted message.
+// This will not call runtime.Caller to get file and line.
+func (e *Error) FastGen(format string, args ...interface{}) error {
+	err := *e
+	err.message = format
+	err.args = args
+	return SuspendStack(&err)
+}
+
+// FastGen generates a new *Error with the same class and code, and a new arguments.
+// This will not call runtime.Caller to get file and line.
+func (e *Error) FastGenByArgs(args ...interface{}) error {
+	err := *e
+	err.args = args
+	return SuspendStack(&err)
+}
+
+// Equal checks if err is equal to e.
+func (e *Error) Equal(err error) bool {
+	originErr := Cause(err)
+	if originErr == nil {
+		return false
+	}
+	if error(e) == originErr {
+		return true
+	}
+	inErr, ok := originErr.(*Error)
+	if !ok {
+		return false
+	}
+	classEquals := e.class.Equal(inErr.class)
+	idEquals := e.ID() == inErr.ID()
+	return classEquals && idEquals
+}
+
+// NotEqual checks if err is not equal to e.
+func (e *Error) NotEqual(err error) bool {
+	return !e.Equal(err)
+}
+
+// ErrorEqual returns a boolean indicating whether err1 is equal to err2.
+func ErrorEqual(err1, err2 error) bool {
+	e1 := Cause(err1)
+	e2 := Cause(err2)
+
+	if e1 == e2 {
+		return true
+	}
+
+	if e1 == nil || e2 == nil {
+		return e1 == e2
+	}
+
+	te1, ok1 := e1.(*Error)
+	te2, ok2 := e2.(*Error)
+	if ok1 && ok2 {
+		return te1.Equal(te2)
+	}
+
+	return e1.Error() == e2.Error()
+}
+
+// ErrorNotEqual returns a boolean indicating whether err1 isn't equal to err2.
+func ErrorNotEqual(err1, err2 error) bool {
+	return !ErrorEqual(err1, err2)
+}
+
+// MarshalJSON implements json.Marshaler interface.
+// aware that this function cannot save a 'registered' status,
+// since we cannot access the registry when unmarshaling,
+// and the original global registry would be removed here.
+// This function is reserved for compatibility.
+func (e *Error) MarshalJSON() ([]byte, error) {
+	return json.Marshal(&struct {
+		Class    ErrClassID  `json:"class"`
```

Comment:
What is class? Is it defined in RFC?

Reply:
https://github.com/pingcap/tidb/blob/master/docs/design/2020-05-08-standardize-error-codes-and-messages.md#the-error-code-range

@overvenus It's mentioned in this section.